### PR TITLE
New Resource: `azurerm_data_protection_backup_instance_mysql_flexible_server`

### PR DESCRIPTION
**Pull request by neil-yechenwei**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description
This PR is to support Data Protection Backup Instance for MySQL FS.

API Reference: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2024-04-01

Overview: https://learn.microsoft.com/en-us/azure/postgresql/flexible-server/concepts-backup-restore#long-term-retention-preview

Note: This PR is referring the existing Backup Instances in RP https://github.com/hashicorp/terraform-provider-azurerm/tree/main/internal/services/dataprotection.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

image
image

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* New Resource: `azurerm_data_protection_backup_instance_mysql_flexible_server`

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: website/docs/r/data_protection_backup_instance_mysql_flexible_server.html.markdown [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+* `name` - (Required) Specifies the name of the Backup Instance for the MySQL Flexible Server. Changing this forces a new resource to be created.
+
+* `location` - (Required) The location of the source database. Changing this forces a new resource to be created.
+
+* `vault_id` - (Required) The ID of the Backup Vault within which the MySQL Flexible Server Backup Instance should exist. Changing this forces a new resource to be created.
+
+* `server_id` - (Required) The ID of the source server. Changing this forces a new resource to be created.
+
+* `backup_policy_id` - (Required) The ID of the Backup Policy.
```
Comment (ms-zhenhua) [inline website/docs/r/data_protection_backup_instance_mysql_flexible_server.html.markdown:93]:
```suggestion
* `backup_policy_id` - (Required) The ID of the Backup Policy.

* `server_id` - (Required) The ID of the source server. Changing this forces a new resource to be created.

* `vault_id` - (Required) The ID of the Backup Vault within which the MySQL Flexible Server Backup Instance should exist. Changing this forces a new resource to be created.
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @neil-yechenwei,
Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

File: internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	return backupinstances.ValidateBackupInstanceID
+}
+
+func (r DataProtectionBackupInstanceMySQLFlexibleServerResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go:53]: Sorry. I didn't find name validation from anywhere.

File: website/docs/r/data_protection_backup_instance_mysql_flexible_server.html.markdown (hunk 1 above)
Comment (neil-yechenwei) [reply website/docs/r/data_protection_backup_instance_mysql_flexible_server.html.markdown]: Updated

File: internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+			existing, err := client.Get(ctx, *id)
+			if err != nil {
+				if response.WasNotFound(existing.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+
+				return fmt.Errorf("reading %s: %+v", *id, err)
+			}
+
+			parameters := *existing.Model
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go:243]: Updated

File: internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+				if props := model.Properties; props != nil {
+					state.Location = location.NormalizeNilable(props.DataSourceInfo.ResourceLocation)
+
+					serverId, err := servers.ParseFlexibleServerID(props.DataSourceInfo.ResourceID)
+					if err != nil {
+						return err
+					}
+					state.ServerId = serverId.ID()
+
+					backupPolicyId, err := backuppolicies.ParseBackupPolicyID(props.PolicyInfo.PolicyId)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go]: Updated

File: internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+			state := BackupInstanceMySQLFlexibleServerModel{
+				Name:    id.BackupInstanceName,
+				VaultId: vaultId.ID(),
+			}
+
+			if model := resp.Model; model != nil {
+				if props := model.Properties; props != nil {
+					state.Location = location.NormalizeNilable(props.DataSourceInfo.ResourceLocation)
+
+					serverId, err := servers.ParseFlexibleServerID(props.DataSourceInfo.ResourceID)
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go]: Updated

File: internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/dataprotection/data_protection_backup_instance_mysql_flexible_server_resource.go]: Updated

**Comment by neil-yechenwei** [comment]
@ms-zhenhua , thanks for the comments. I updated PR. Please take another look.

**Review by ms-zhenhua — APPROVED** [review]

**Comment by ms-zhenhua** [comment]
Hi @neil-yechenwei , thank you for your update. LGTM~

**Review by katbyte — CHANGES_REQUESTED** [review]
test failure
```
------- Stdout: -------
=== RUN   TestAccDataProtectionBackupInstanceMySQLFlexibleServer_basic
=== PAUSE TestAccDataProtectionBackupInstanceMySQLFlexibleServer_basic
=== CONT  TestAccDataProtectionBackupInstanceMySQLFlexibleServer_basic
    testcase.go:173: Step 1/3 error: Error running apply: exit status 1
        
        Error: creating Flexible Server (Subscription: "*******"
        Resource Group Name: "acctest-dataprotection-241016175940838067"
        Flexible Server Name: "acctest-mysqlfs-241016175940838067"): polling after Create: polling failed: the Azure API returned the following error:
        
        Status: "ZoneNotAvailableForRegion"
        Code: ""
        Message: "The requested size for resource is currently not available in this zone. Please try another zone or deploy to a different location"
        Activity Id: ""
        
        ---
        
        API Response:
        
        ----[start]----
        {"name":"d4b8a22a-f57b-4863-965a-cf540392eb18","status":"Failed","startTime":"2024-10-16T18:00:08.933Z","error":{"code":"ZoneNotAvailableForRegion","message":"The requested size for resource is currently not available in this zone. Please try another zone or deploy to a different location"}}
        -----[end]-----
        
        
          with azurerm_mysql_flexible_server.test,
          on terraform_plugin_test.tf line 37, in resource "azurerm_mysql_flexible_server" "test":
          37: resource "azurerm_mysql_flexible_server" "test" {
        
--- FAIL: TestAccDataProtectionBackupInstanceMySQLFlexibleServer_basic (223.58s)
FAIL
```

**Comment by neil-yechenwei** [comment]
@katbyte , it's the quota issue on Teamcity. It's not related with this PR and it passed on my local.

Below is the test result that I just now triggered.
```
--- PASS: TestAccDataProtectionBackupInstanceMySQLFlexibleServer_basic (822.97s)
```

**Comment by katbyte** [comment]
@neil-yechenwei - went to test this and the build failed for tc due to an API upgrade on main

```
 data_protection_backup_instance_mysql_flexible_server_resource.go:17:2: cannot find module providing package github.com/hashicorp/go-azure-sdk/resource-manager/mysql/2022-01-01/servers: import lookup disabled by -mod=vendor
[08:57:37][Step 4/5] 	(Go version in go.mod is at least 1.14 and vendor directory exists.)
```

you'll need to merge main and resolve conflicts

**Comment by neil-yechenwei** [comment]
@katbyte , I merged latest change. Below is the latest test result I just now triggered.

image
image

Note: the failure of CI check is not related with this PR.

**Pull request merged**